### PR TITLE
[RISCV][MC] Support Assembling 48- and 64-bit Instructions

### PR DESCRIPTION
**Pull request by lenary**
This adds `.insn` support for assembling instructions of 48- and 64-bits. Disassembly already knows to bunch up the instruction bits for these instructions.

This changes some error messages so they are a little clearer.

---

I didn't go beyond 64-bit as that requires significant surgery to the backend's instruction encoding mechanism, whereas we can support up to 64-bit with these minimal changes here.

### CODE REVIEW

**Comment by jrtc27** [comment]
As far as I know the encoding for >32-bit instructions is not ratified, nor even stable/frozen. Has that changed recently?

**Comment by jrtc27** [comment]
https://github.com/riscv/riscv-isa-manual/blob/7023c601443adf21850772e02e3d920bc68b2237/src/intro.adoc?plain=1#L496 seems to still be there

**Comment by lenary** [comment]
Is that the only documentation for the fact they're not ratified? Note the preface says "Standard instruction-length encodings have been defined for 48-bit, 64-bit, and 64-bit
instructions.", and Table 70 (which may be deemed not to be the canonical encoding information) also lacks that note.

This patch follows Table 70, and kicks the can down the road for much longer instructions, but maybe that still means there's no path to landing this and supporting hand-assembling wider instructions so they can be evaluated?

Alternatively, is it maybe worth adding a feature that we can put behind `-menable-experimental-extensions` so LLVM can support this without the wider encodings needing to be ratified?

**Comment by jrtc27** [comment]
That addition came out of a bunch of discussion in https://github.com/riscv/riscv-isa-manual/issues/280 and resulted in https://github.com/riscv/riscv-isa-manual/commit/1d8fbc86a56d9b200bdc6fa5d20afa2aa0d3af8d, which post-dates 2.0, the description of which is where the quote you took comes from. "Table 70. RISC-V base opcode map, inst[1:0]=11" would benefit from clarifying that the encodings are reserved but not ratified for that specific use.

**Comment by lenary** [comment]
Ok, I will revisit this PR tomorrow, to put these behind an experimental extension. I'm much less clear on what should be done in the disassembler though, as it also uses this encoding information.

**Comment by lenary** [comment]
Looking back on the patch this morning, the only thing that is actually dependent on the length encodings themselves is actually the length detection logic, so I have updated the patch to never attempt to understand if an instruction might be longer than 4 bytes - instead you have to provide an explicit length for 6/8 byte encodings. I think this is a much better way forwards than a new feature, while the encodings are not yet frozen - and should still allow experimentation.

**Comment by lenary** [comment]
I hope this is ready for a re-review, given I think I have adequately taken into account the feedback from @jrtc27 about the instability of longer encodings, and have hopefully found a way forwards to allow users to begin experimenting with longer instruction encodings without being beholden to something that is not yet frozen/ratified in the specification.

I think once the specification does freeze/ratify some encodings for longer instructions, we can maybe start inferring lengths from the encoding again, which is to say I don't think this patch's behaviour has to be how `.insn` works indefinitely into the future.

I just added some docs to the review as well, to document how to use `.insn` for longer instructions, given the behaviour is no longer straightforward.

**Comment by dtcxzyw** [comment]
> I hope this is ready for a re-review, given I think I have adequately taken into account the feedback from @jrtc27 about the instability of longer encodings, and have hopefully found a way forwards to allow users to begin experimenting with longer instruction encodings without being beholden to something that is not yet frozen/ratified in the specification.
> 
> I think once the specification does freeze/ratify some encodings for longer instructions, we can maybe start inferring lengths from the encoding again, which is to say I don't think this patch's behaviour has to be how `.insn` works indefinitely into the future.
> 
> I just added some docs to the review as well, to document how to use `.insn` for longer instructions, given the behaviour is no longer straightforward.
> 
> PS: I will be away for a week, so not able to respond to reviews.

FYI binutils supports inferring the length from its encoding: https://github.com/bminor/binutils-gdb/blob/428f3561bc16dfc2944ee641201acdd166315aa2/gas/testsuite/gas/riscv/insn.s#L74-L81. But I agree that we can improve this after the spec is ratified :)

**Review by topperc — APPROVED** [review]
LGTM

File: llvm/docs/RISCVUsage.rst [hunk 1]
```diff
@@ -426,6 +426,20 @@ line.  This currently applies to the following extensions:
 
 No extensions have experimental intrinsics.
 
+Long (>32-bit) Instruction Support
+==================================
+
+RISC-V is a variable-length ISA, but the standard currently only defines 16- and 32-bit instructions. The specification describes longer instruction encodings, but these are not stable.
```
Comment (asb) [inline llvm/docs/RISCVUsage.rst:432]: are not stable => not ratified

We use "ratified" a lot elsewhere in the document, but not "stable" at all so I think it's the better term to use.

File: llvm/docs/RISCVUsage.rst [hunk 2]
```diff
@@ -426,6 +426,20 @@ line.  This currently applies to the following extensions:
...
+
+* ``.insn type, args*`` which takes a known instruction type, and a list of fields. You are strongly recommended to use this variant of the directive if your instruction fits an existing instruction type.
+* ``.insn [ length , ] encoding`` which takes an (optional) explicit length (in bytes) and a raw encoding for the instruction. When given an explicit length, this variant can encode instructions up to 64 bits long. The encoding part of the directive must be given all bits for the instruction, none are filled in for the user. When used without the optional length, this variant of the directive will use the LSBs of the raw encoding to work out if an instruction is 16 or 32 bits long. LLVM does not infer that an instruction might be longer than 32 bits - in this case, the user must give the length explicitly.
+
+It is stringly recommended to use the ``.insn`` directive for assembling unsupported instructions instead of ``.word`` or ``.hword``, because it will produce the correct mapping symbols to mark the word as an instruction, not data.
```
Comment (asb) [inline llvm/docs/RISCVUsage.rst:441]: stringly => strongly

File: llvm/test/MC/RISCV/insn-invalid.s [hunk 3]
```diff
@@ -26,8 +26,28 @@
...
+.insn 0x6, 0x000000000001 # CHECK: :[[@LINE]]:7: error: compressed instructions are not allowed
+.insn 0x8, 0x0000000000000001 # CHECK: :[[@LINE]]:7: error: compressed instructions are not allowed
+
+.insn 0x2, 0x10001 # CHECK: :[[@LINE]]:7: error: encoding value does not fit into instruction
+.insn 0x4, 0x100000003 # CHECK: :[[@LINE]]:7: error: encoding value does not fit into instruction
+.insn 0x6, 0x100000000001f # CHECK: :[[@LINE]]:7: error: encoding value does not fit into instruction
+.insn 0x8, 0x1000000000000003f # CHECK: :[[@LINE]]:12: error: expected an integer constant
+
+.insn 0x0010 # CHECK: :[[@LINE]]:7: error: compressed instructions are not allowed
+.insn 0x2, 0x0001 # CHECK: :[[@LINE]]:7: error: compressed instructions are not allowed
```
Comment (asb) [inline llvm/test/MC/RISCV/insn-invalid.s:53]: I don't think you have any test coverage for specifying a length that isn't a multiple of two bytes.

**Review by asb — APPROVED** [review]
Left some minor comments but this seems good to me. I think this is a good solution to the issue of the longer encodings not being ratified, so thanks for finding it!

**Comment by lenary** [comment]
@asb I think I've done all the requested changes. I'll wait until tomorrow to merge.

**Pull request merged**